Source/DafnyCore/Resolver/ExpectContracts.cs

```diff
@@ -0,0 +1,291 @@
+
+public class ExpectContracts : IRewriter {
+  private readonly ClonerButDropMethodBodies cloner = new();
+  private readonly Dictionary<MemberDecl, MemberDecl> wrappedDeclarations = new();
+  private readonly Dictionary<MemberDecl, string> fullNames = new();
+  private readonly Dictionary<string, MemberDecl> newDeclarationsByName = new();
+  private CallRedirector callRedirector;
+
+  public ExpectContracts(ErrorReporter reporter) : base(reporter) {
+    callRedirector = new(reporter);
+  }
+
+  /// <summary>
+  /// Create an expect statement that checks the given contract clause
+  /// expression and fails with a message that points to the original
+  /// location of the contract clause if it is not true at runtime.
+  ///
+  /// If the given clause is not compilable, emit a warning and construct
+  /// an `expect true` statement with a message explaining the situation.
+  /// </summary>
+  /// <param name="expr">The contract clause expression to evaluate.</param>
+  /// <param name="exprType">Either "requires" or "ensures", to use in the
+  /// failure message.</param>
+  /// <returns>The newly-created expect statement.</returns>
+  private Statement CreateContractExpectStatement(AttributedExpression expr, string exprType) {
+    var tok = expr.E.tok;
+    var msg = $"Runtime failure of {exprType} clause from {tok.filename}:{tok.line}:{tok.col}";
+    var exprToCheck = expr.E;
+    if (ExpressionTester.UsesSpecFeatures(exprToCheck)) {
+      Reporter.Warning(MessageSource.Rewriter, tok,
+        $"The {exprType} clause at this location cannot be compiled to be tested at runtime because it references ghost state.");
+      exprToCheck = new LiteralExpr(tok, true);
+      msg += " (not compiled because it references ghost state)";
+    }
+    var msgExpr = Expression.CreateStringLiteral(tok, msg);
+    return new ExpectStmt(tok, expr.E.EndToken, exprToCheck, msgExpr, null);
+  }
+
+  /// <summary>
+  /// Creates a block statement that includes an expect statement for every
+  /// requires clause, followed by the provided call statement, followed by
+  /// an expect statement for every ensures clause.
+  /// </summary>
+  /// <param name="requires">The list of requires clause expressions.</param>
+  /// <param name="ensures">The list of ensures clause expressions.</param>
+  /// <param name="callStmt">The call statement to include.</param>
+  /// <returns>The newly-created block statement.</returns>
+  private BlockStmt MakeContractCheckingBody(IEnumerable<AttributedExpression> requires,
+    IEnumerable<AttributedExpression> ensures, Statement callStmt) {
+    var expectRequiresStmts = requires.Select(req =>
+      CreateContractExpectStatement(req, "requires"));
+    var expectEnsuresStmts = ensures.Select(ens =>
+      CreateContractExpectStatement(ens, "ensures"));
+    var callStmtList = new List<Statement>() { callStmt };
+    var bodyStatements = expectRequiresStmts.Concat(callStmtList).Concat(expectEnsuresStmts);
```

Comment:
That's interesting. So your approach is to rewrite any extern method call to inline the requires call and the expects clause.
Do you need to check the requires? Isn't that going to be checked by Dafny statically?
For the ensures, it's great, because it means one can write a test fuzzer without any expect, and the method's postcondition will still be checked. Could you please add this idea in documentation of /testContracts?

Reply:
For externs (of the form that this is focused on, where the implementation is written in the target language), that's true. Dafny will prove that the requires clause is satisfied. However, if you were to apply a fuzzer to the target language code, the existence of a branch on a requires clause is likely to make it easier to come up with inputs that satisfy that requires clause. Arguably, perhaps, it might make sense to check all of the requires clauses together in one `expect` statement, but check the ensures clauses separately, as the current code does.

I think any changes like that would be driven by how well it works in practice, though.

I'll add some reference manual documentation of `/testContracts` that goes into this.

Reply:
> However, if you were to apply a fuzzer to the target language code, the existence of a branch on a requires clause is likely to make it easier to come up with inputs that satisfy that requires clause.

If you were to apply a fuzzer to the target language code, should you not gracefully exit if the requires is not satisfied?

Reply:
Ideally, yes, though I think that the current structure would at least give you enough to get started. If you run a fuzzer, and filter out the cases of failing requires clauses, you'll get a useful outcome.

In the long run, I'd like this code to integrate with target language fuzzing/verification APIs, using `assume`-like things for `requires` clauses and `assert`-like things for `ensures` clauses. I think that's a bit down the road, though.